src/html.cr

```diff
@@ -36,10 +36,33 @@ module HTML
   # HTML.escape("Crystal & You", io) # => nil
   # io.to_s                          # => "Crystal &amp; You"
   # ```
+  @[AlwaysInline]
   def self.escape(string : String, io : IO) : Nil
-    string.each_char do |char|
-      io << SUBSTITUTIONS.fetch(char, char)
+    escape(string.to_slice, io)
+  end
+
+  # Same as `escape(String, IO)` but accepts `Bytes` instead of `String`.
+  #
+  # WARNING: The slice is assumed to be valid UTF-8.
+  def self.escape(string : Bytes, io : IO) : Nil
+    last_copy_at : Int32 = 0
+    string.each_with_index do |byte, index|
+      str = case byte
+            when '&'  then "&amp;"
```

Comment:
Yes, most of the performance improvement should come from avoiding the hash lookups.
That duplication shouldn't be much of a problem. It would make sense to add comments with references between both locations.
But perhaps we can even merge them. Technically, we could create the `when` branches from `SUBSTITUTIONS` via macro expansion. I don't think that would be wise, though. It only increases complexity for a very small problem.

The implementation of `escape(String)` could just delegate to `escape(String, io)` which makes the constant unused (it's private, so we can simply remove it).

The proposal in https://github.com/crystal-lang/crystal/pull/13139#discussion_r1123191440 could be another option.

Reply:
What happens if `SUBSTITUTIONS` were a `NamedTuple`?

Reply:
Yeah `unsafe_chr.to_s` allocates a string for every character. Of course that's inefficient. I'm actually suprised the benchmark isn't that much worse 😄 (I might suspect some kind of crazy LLVM optimizations at work)